jax/experimental/export/shape_poly.py

```diff
@@ -1033,17 +1036,31 @@ def shape(self, tok: tokenize.TokenInfo) -> tuple[Sequence[DimSize], tokenize.To
     while True:
       if tok.exact_type in self.FOLLOW_SHAPE:
         break
+      # Error checking in presence of placeholders
+      if (tok.exact_type == tokenize.ELLIPSIS or
+          tok.exact_type == tokenize.NAME and tok.string == "_"):
+        if self.like_shape is None:
+          raise self.parse_err(tok,
+                               "spec contains ... but no 'like' shape was given")
+        if tok.exact_type == tokenize.ELLIPSIS:
+          min_len_like_shape = len(self.dimensions)
+        else:
+          min_len_like_shape = len(self.dimensions) + 1
+        if len(self.like_shape) < min_len_like_shape:
+          raise self.parse_err(
+            tok,
+            f"cannot resolve placeholder '{tok.string}' because we parsed "
+            f"already {len(self.dimensions)} and 'like' shape has "
+            f"only {len(self.like_shape)} dimensions")
       if tok.exact_type == tokenize.ELLIPSIS:
-        to_add = self.arg_shape[len(self.dimensions):]
+        to_add = self.like_shape[len(self.dimensions):]  # type: ignore
```

Comment:
OOC what doesn't mypy like here and below? Consider ignoring a specific error: `type: ignore[foo-error]`.